### PR TITLE
Bring `d3d12` crate in-tree

### PR DESCRIPTION
**Pull request by ErichDonGubler**
I've [noted](https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$Fv-F0YkeGQcfcLE_cpuwmrqnjDs2qVIBANHHD8mU2BY?via=matrix.org&via=mozilla.org&via=kde.org) in Matrix chat that `d3d12`, as a crate, sits in an awkward place. It does not realize the potential it was apparently envisioned to have of being a robust abstraction that others can use, like with [`metal`](https://github.com/gfx-rs/metal-rs) or [`ash`](https://github.com/ash-rs/ash). It remains totally coupled to WGPU's needs. And yet, `d3d12` also reaps no benefits of its coupling. Many D3D12 APIs used WGPU's D3D backends are bound only in `wgpu-hal`, which defeats the purpose of having a dedicated layer for D3D12 interop. For those APIs that _are_ captured by the `d3d12` crate, effective development requires two compilation feedback loops _across repositories_. ~~The crate barrier also inhibits Clippy and `rustc` lints that would surely be useful for code that could otherwise be a module tree in `wgpu-hal`.~~

@cwfitzgerald [noted](https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$2zTYNx4FmVJr_GDexw5QP3mCGZgG5UIHNB84fHl_9FM?via=matrix.org&via=mozilla.org&via=kde.org) that he'd be open to inlining `d3d12` code into `wgpu-hal`. I also took time to get some Mozilla consensus; @jimblandy laughed in apparent catharsis when I declared that `d3d12` is, proverbially, a child who either needs to get a job and pay its rent, or move back in with its parents (i.e., `wgpu-hal`). So, here's the PR making moving the entirety of the `d3d12` crate from there to here!

**This PR copies the subtree-merges `d3d12` repo into a new `d3d12` top-level folder.** ~~It also fixes up some integration issues that appear unique to the current `HEAD` of the `d3d12` repo.~~

Note that while this is inlining `d3d12` _for now_, there's no reason that we can't or shouldn't eventually try to grow the in-tree crate into a well-formed separate repository again. No plans for that are established here, but it's a definite future possibility.

---

~~I've taken pains to ensure that actual changes to code are easy to distinguish as small commits, while keeping mechanical movement of code to larger ones. Therefore, **recommended review experience is by commit**.~~

Post-merge checklist:

- [ ] Archive the `d3d12` crate?

---

**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**
_Link to the issues addressed by this PR, or dependent PRs in other repositories_

Linked in prose above.

**Description**
_Describe what problem this is solving, and how it's solved._

See my prose above.

**Testing**
_Explain how this change is tested._

No behavior should change because of this PR. I consider successful compilation to be a sufficient test for it.

### CODE REVIEW

File: wgpu-hal/Cargo.toml
Comment (ErichDonGubler) [inline wgpu-hal/Cargo.toml]: TODO: Split out putting implied features into their own lines in a preparatory commit.

File: wgpu-hal/src/auxil/dxgi/bindings/mod.rs
Comment (ErichDonGubler) [inline wgpu-hal/src/auxil/dxgi/bindings/mod.rs:1]: TODO: There's several `clippy` lints that need to be resolved here before integration. They don't look too hard, but they should probably be separate commits prior to the one moving `d3d12` code into `wgpu-hal`.

**Comment by teoxoy** [comment]
I think there is still value in keeping it as a separate crate. I thought the plan was to only move it in this repository so that we can iterate on it faster. Once we are happy with it, we can publish it as well.

**Comment by ErichDonGubler** [comment]
@teoxoy:

> I think there is still value in keeping it as a separate crate. I thought the plan was to only move it in this repository so that we can iterate on it faster. Once we are happy with it, we can publish it as well.

Agreed. I think that eventually publishing well-formed D3D* bindings are well within the purview of this org. Once we have the resources to truly create it, we should do it. I've added this to the OP:

> Note that while this is inlining `d3d12` _for now_, there's no reason that we can't or shouldn't eventually try to grow the inlined code into a well-formed separate crate again. No plans for that are established here, but it's a definite future possibility.

**Comment by cwfitzgerald** [comment]
Following up on this, what's the current deal?

**Comment by cwfitzgerald** [comment]
Whoops wrong button.

**Comment by ErichDonGubler** [comment]
Short answer: will get to this when there's some more impetus from the `wgpu` community or Firefox's side. Firefox has a few bugs in our triage queue that may quickly turn this into a priority, but it's not currently.

**Comment by ErichDonGubler** [comment]
Just to ensure we have consensus: @cwfitzgerald, @jimblandy, and I have been thinking that `d3d12` should be inlined as a set of modules within `wgpu-hal`. @teoxoy, are you saying that this is acceptable iff we plan on eventually breaking it back out again into a fully formed crate? Or did you not want to go past the point of inlining `d3d12`-the-crate into the same Git repository?

**Comment by cwfitzgerald** [comment]
The plan has always ended with a vague "once things are settled evaluate if it should be split back out" at the end

**Comment by ErichDonGubler** [comment]
Actually, since it's relevant to the question I've posed of consensus, _and_ I think it'll be helpful to have multiple stages for this inlining, I'll limit scope in this PR. Step two would be to do the inlining anyway; there are a lot of lints to cover when we actually convert `d3d12` into modules, assuming we reach consensus with @teoxoy.

File: wgpu-hal/Cargo.toml
Comment (ErichDonGubler) [reply wgpu-hal/Cargo.toml]: Outdated, gonna move to a different PR.

File: wgpu-hal/src/auxil/dxgi/bindings/mod.rs
Comment (ErichDonGubler) [reply wgpu-hal/src/auxil/dxgi/bindings/mod.rs:1]: Outdated, gonna move to a different PR.

**Comment by ErichDonGubler** [comment]
I'm getting failures for `wasm32` tests, since CI is attempting to compile `d3d12` for the WASM target. Not sure how to fix this yet, but I'll drill down into it once Windows failures are cleared up (we're getting several, ATM).

**Comment by cwfitzgerald** [comment]
We probably need a `#![cfg(not(target_arch = "wasm32"))]` at the top level

**Comment by ErichDonGubler** [comment]
@cwfitzgerald: Figured it out! We want to `exclude` `d3d12` in the workspace `Cargo.toml`, because [`path` dependencies are counted as `members` by default](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields):

> All [`path` dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies) residing in the workspace directory automatically become members. …
>
> …
>
> The `exclude` key can be used to prevent paths from being included in a workspace. This can be useful if some path dependencies aren’t desired to be in the workspace at all, or using a glob pattern and you want to remove a directory.
>
> …

Fixed with a419a50e.

**Comment by cwfitzgerald** [comment]
Hmm interesting - just verify clippy, tests, etc still work with it.

**Comment by teoxoy** [comment]
My main concern at the time was that we'll be "inlining" all `d3d12-rs`'s crate code into our d3d12 backend. The PR was still WIP and looking at what the PR is doing wasn't enough to gauge its end state. My interpretation of  "inlining" being the traditional ["replaces a function call site with the body of the called function"](https://en.wikipedia.org/wiki/Inline_expansion).

After talking about it in a meeting with @ErichDonGubler he mentioned that wasn't the case. We would just keep the contents of the crate in a module.

This then became a question of should this be a module or a crate?

I felt like the answer was: keep it as a crate since it's already one and having to convert it to a module to then convert it back once the status quo improves felt unnecessary.

This is however not a strong opinion.

**Comment by cwfitzgerald** [comment]
> I felt like the answer was: keep it as a crate since it's already one and having to convert it to a module to then convert it back once the status quo improves felt unnecessary.

Ah yeah, leaving it a crate seems fine. It being in-tree is the most important part.

**Comment by ErichDonGublerUnprivileged** [comment]
@gfx-rs/wgpu

**Comment by ErichDonGubler** [comment]
@cwfitzgerald (CC @teoxoy, @jimblandy): Interestingly, I kept running into crashes and errors regarding certain COM objects apparently being null after naively switching to the in-tree copy of `d3d12` ([1](https://github.com/gfx-rs/wgpu/actions/runs/6253114430/job/16977821472?pr=3987), [2](https://github.com/gfx-rs/wgpu/actions/runs/6254174102/job/16981148682?pr=3987)). I resolved this by removing the nullability of `ComPtr`, but I don't think the current set of internal APIs is great. There's a _lot_ of places where code apparently tries to be a "thin" FFI into Windows, but ends up missing opportunities to eliminate accidental complexity in our usage of DX.

I think it'd be much better to transition to using stronger `Result` types, perhaps as a follow-up PR. I've gotten everything compiling and passing again here; I'm planning on splitting commits with the current end result of code. Would you like to see any further major end-state changes before I go to the effort?

**Comment by ErichDonGubler** [comment]
@cwfitzgerald, @teoxoy: I have updated the OP to indicate the current plan to just leave the in-tree copy of `d3d12` at that, and not inline any code into modules in `wgpu-hal`. Thanks!

**Review by teoxoy — DISMISSED** [review]
Good stuff!

> I think it'd be much better to transition to using stronger `Result` types, perhaps as a follow-up PR. Would you like to see any further major end-state changes before I go to the effort?

Sounds great; anything getting us closer to more idiomatic rust I'm rooting for!

**Comment by ErichDonGubler** [comment]
@teoxoy: Marked as `Ready for review`, with only these changes (`git diff 652ad50..d31c3ab0 -- d3d12/`):

```diff
diff --git a/d3d12/src/com.rs b/d3d12/src/com.rs
index d137f120..db790e8a 100644
--- a/d3d12/src/com.rs
+++ b/d3d12/src/com.rs
@@ -10,14 +10,12 @@ use winapi::{um::unknwnbase::IUnknown, Interface};
 pub struct ComPtr<T: Interface>(*mut T);
 
 impl<T: Interface> ComPtr<T> {
-    // TODO: remove `mut_void`, null`/`is_null`
-
     /// Create a ComPtr from a raw pointer. This will _not_ call AddRef on the pointer, assuming
     /// that it has already been called.
     ///
     /// # Safety
     ///
-    /// - if `raw` is not null, it must be a valid pointer to a COM object that implements T.
+    /// - `raw` must be a valid pointer to a COM object that implements T.
     pub unsafe fn from_reffed(raw: *mut T) -> Self {
         debug_assert!(!raw.is_null());
         ComPtr(raw)
@@ -34,29 +32,15 @@ impl<T: Interface> ComPtr<T> {
         ComPtr(raw)
     }
 
-    /// Returns the raw inner pointer. May be null.
+    /// Returns the raw inner pointer.
     pub fn as_ptr(&self) -> *const T {
         self.0
     }
 
-    /// Returns the raw inner pointer as mutable. May be null.
+    /// Returns the raw inner pointer as mutable.
     pub fn as_mut_ptr(&self) -> *mut T {
         self.0
     }
-
-    /// Returns a mutable reference to the inner pointer.
-    ///
-    /// This is useful when D3D functions initialize objects by filling in a pointer to pointer
-    /// by taking `T**` as an argument.
-    ///
-    /// # Safety
-    ///
-    /// - Any modifications done to this pointer must result in the pointer either:
-    ///   - being set to null
-    ///   - being set to a valid pointer to a COM object that implements T
-    pub fn mut_self(&mut self) -> &mut *mut T {
-        &mut self.0
-    }
 }
 
 impl<T: Interface> ComPtr<T> {
```

If I don't see an update in the next day or so, I'll assume that you're okay with what is effectively doc-comment clean-up, and just merge.

EDIT: Wrong commit, sorry! 😖

**Comment by ErichDonGubler** [comment]
TIL GitHub added the `Compare` control in-line with force push events show in this feed. 👀

**Comment by cwfitzgerald** [comment]
It was always there, just completely non-obvious. If you click on "force-pushed" you'll see a diff too.

File: d3d12/.github/workflows/ci.yml
Comment (cwfitzgerald) [inline d3d12/.github/workflows/ci.yml]: This shouldn't be included

File: d3d12/.gitignore
Comment (cwfitzgerald) [inline d3d12/.gitignore:1]: Same hjere

File: d3d12/rustfmt.toml
Comment (cwfitzgerald) [inline d3d12/rustfmt.toml:1]: This file shouldn't be here, especially empty

**Review by cwfitzgerald — DISMISSED** [review]

File: d3d12/.github/workflows/ci.yml
Comment (ErichDonGubler) [reply d3d12/.github/workflows/ci.yml]: I think cribbing CI from `wgpu` would be an overall improvement, so I see not reason not to do this for now. 🙂

Done with 6d355881.

File: d3d12/rustfmt.toml
Comment (ErichDonGubler) [reply d3d12/rustfmt.toml:1]: Removed with 0614445a.

File: d3d12/.gitignore
Comment (ErichDonGubler) [reply d3d12/.gitignore:1]: Removing this actually seems undesirable, since this directory is its technically own crate outside the main workspace. We happen to still ignore it because of how the `.gitignore` matching rules work (i.e., `target/` in our top-level `.gitignore` matches a `target` folder in _all_ subdirectories 😶), but this `.gitignore` still makes sense, IMO.

I'll still delete it if you want, but I don't think this is the same class of redundancy as the `rustfmt.toml` or GH actions configuration.

**Comment by ErichDonGubler** [comment]
@teoxoy, @cwfitzgerald: I have reverted to draft and marked this PR as ready again, re-requesting your reviews. At @kdashg's encouragement, I decided to actually try to diagnose apparent issues with `ComPtr::mut_void`. This has resulted in significant changes to this current PR, making the diff. overall much less complex.

The issues with testing appear to have been a regression introduced in https://github.com/gfx-rs/d3d12-rs/pull/50 (authored by @cwfitzgerald), which was not released or consumed here before. A summary of the problems we encountered can be found [here](https://github.com/gfx-rs/d3d12-rs/pull/50#discussion_r1337512140). I've fixed in this PR with 1ad40cf8.

File: d3d12/src/com.rs
Comment (kdashg) [inline d3d12/src/com.rs]: Why use cast? I think we shouldn't.
It is risky to change anything vs what we know worked before.

**Review by kdashg — CHANGES_REQUESTED** [review]

**Comment by cwfitzgerald** [comment]
I think we should split this change - one MR  to bring in exactly what we depended on previously. Then follow up MRs to improve/iterate on the apis.

**Comment by kdashg** [comment]
> I think we should split this change - one MR to bring in exactly what we depended on previously. Then follow up MRs to improve/iterate on the apis.

Thank you, this has been my pitch as well.

**Comment by ErichDonGubler** [comment]
I will plan on changing this PR to “just” bring 0.7 (or a fixed `d3d12` upstream) in-tree.

**Comment by kdashg** [comment]
Specifically:
1. revert gfx-rs/d3d12-rs#50
2. Optional: Re-fix the parts of gfx-rs/d3d12-rs#50 that don't touch ComPtr.
3. Required: Leave ComPtr untouched
4. Required: Test this new d3d12-rs with Firefox.
  4a. Ensure there are no changes to https://searchfox.org/mozilla-central/rev/57f6fbd39c0b5957e11b27b4db58b821d8e1607d/third_party/rust/d3d12/src/com.rs#37
6. Inline d3d12-rs into wgpu
7. Required: Test this new wgpu with Firefox.
6. Optional: Further improvements

**Comment by cwfitzgerald** [comment]
I think we should leave d3d12 completely untouched from the commit we were depending on. That way we can land the organizational change without testing on Firefox, and reducing the pressure to land all of the subsequent changes

**Comment by kdashg** [comment]
I'm also happy to do that, what I've been thinking of as "the most safe option".

**Comment by ErichDonGubler** [comment]
FTR: the older discussion about inlining code from `d3d12` into `wgpu-hal` can be found in Matrix [here](https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$2zTYNx4FmVJr_GDexw5QP3mCGZgG5UIHNB84fHl_9FM?via=matrix.org&via=mozilla.org&via=kde.org).

**Review by cwfitzgerald — DISMISSED** [review]
Dismissing my own review for bookkeeping.

File: Cargo.toml
Comment (cwfitzgerald) [inline Cargo.toml:21]: Why are we excluding d3d, this should be a default member of the repo?

File: d3d12/bors.toml
Comment (cwfitzgerald) [inline d3d12/bors.toml:1]: This file can definitely be nuked

File: d3d12/rustfmt.toml
Comment (cwfitzgerald) [reply d3d12/rustfmt.toml:1]: Seems to still be here

File: wgpu-hal/Cargo.toml
Comment (cwfitzgerald) [inline wgpu-hal/Cargo.toml:141]: This should both be a path and a version I think, to let us publish the crate without changing this.

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]

File: Cargo.toml
Comment (ErichDonGubler) [reply Cargo.toml:21]: Per previous conversation [here](https://github.com/gfx-rs/wgpu/pull/3987#issuecomment-1728471456), this is to avoid breakage with WASM builds (since all members of the workspace are attempted to be built, but we don't actually want that in WASM builds).

File: wgpu-hal/Cargo.toml
Comment (ErichDonGubler) [reply wgpu-hal/Cargo.toml:141]: Fixed up with 22d01e452.

File: d3d12/bors.toml
Comment (ErichDonGubler) [reply d3d12/bors.toml:1]: This sounds like an adjustment to make after merge, yes? As you [stated earlier](https://github.com/gfx-rs/wgpu/pull/3987#issuecomment-1736213961), our intent here is to bring in the original source as-is, unless it's breaking something, with refinements in follow-up PRs.

File: Cargo.toml
Comment (ErichDonGubler) [inline Cargo.toml:178]: @cwfitzgerald: FYI, added this in with 887868927.

File: d3d12/bors.toml
Comment (cwfitzgerald) [reply d3d12/bors.toml:1]: Ah alright, If I said that, do that :)

File: Cargo.toml
Comment (cwfitzgerald) [reply Cargo.toml:21]: Okay lets fix this in a follow up

**Review by cwfitzgerald — APPROVED** [review]

**Pull request merged**
